tracing-subscriber/src/fmt/format.rs

```diff
@@ -280,8 +309,54 @@ where
     }
 }
 
-// === impl FormatFields ===
+#[cfg(feature = "json")]
+impl<N, T> FormatEvent<N> for Format<Json, T>
+where
+    N: for<'writer> FormatFields<'writer>,
+    T: FormatTime,
+{
+    fn format_event(
+        &self,
+        ctx: &span::Context<'_, N>,
+        writer: &mut dyn fmt::Write,
+        event: &Event<'_>,
+    ) -> fmt::Result {
+        let mut timestamp = String::new();
+        self.timer.format_time(&mut timestamp)?;
+
+        #[cfg(feature = "tracing-log")]
+        let normalized_meta = event.normalized_metadata();
+        #[cfg(feature = "tracing-log")]
+        let meta = normalized_meta.as_ref().unwrap_or_else(|| event.metadata());
+        #[cfg(not(feature = "tracing-log"))]
+        let meta = event.metadata();
+
+        let mut visit = || {
+            let mut serializer = Serializer::new(WriteAdaptor::new(writer));
+            let mut serializer = serializer.serialize_map(None)?;
+
+            serializer.serialize_entry("timestamp", &timestamp)?;
+            serializer.serialize_entry("level", &meta.level().as_serde())?;
+
+            ctx.with_current(|(_, span)| serializer.serialize_entry("span", span.name()))
```

Comment:
this will only serialize the name of the immediate parent span. it seems like it might be good to serialize the span's fields as well — note that if we don't also serialize fields, the JSON formatter will output strictly less information than the text formatter, which does include the parent span's fields.

Reply:
The current `Span` implementation in `tracing-subscriber` formats fields in a format suitable for plain test logs, but looks kind of bad in JSON. See the test example here:

https://github.com/pimeys/tracing/commit/b143824b9e1390f79527e7e680a00f324ea9046c#diff-12298a6463a49769f9b40d8e8b01e754R817

I couldn't really track where and why do we stringify the fields in this way. Would be nicer to have a struct that implements `Display` and `AsSerde` to the both formats.

Reply:
The fields are formatted by a type implementing the `FormatFields` trait:
https://github.com/tokio-rs/tracing/blob/06daa9512b250e2d73e8722f7f15dfcb9d95f3be/tracing-subscriber/src/fmt/format.rs#L49-L64
The `FormatFields` implementation used can be set by the `Builder` here:
https://github.com/tokio-rs/tracing/blob/06daa9512b250e2d73e8722f7f15dfcb9d95f3be/tracing-subscriber/src/fmt/mod.rs#L402-L433
`FormatFields` is also implemented for any `MakeVisitor` implementation:
https://github.com/tokio-rs/tracing/blob/06daa9512b250e2d73e8722f7f15dfcb9d95f3be/tracing-subscriber/src/fmt/format.rs#L285-L299

So, we could override the default field formatter to record the fields using a `Serde` visitor. Although the formatted representation would be stored as a string, we could then interpolate that string into the output, and the fields would be formatted as JSON.

Reply:
<3